bundles/org.openhab.binding.mycroft/README.md

```diff
@@ -0,0 +1,161 @@
+
+| property      |  type                           | description                                                             | mandatory |
+|---------------|---------------------------------|-------------------------------------------------------------------------|-----------|
+| messageTypes  | List of string, comma separated | Only these message types will be forwarded to the Full Message Channel  |   No      |
+
+
+## Full Example
+
+A manual setup through a `things/mycroft.things` file could look like this:
+
+```java
+Thing mycroft:mycroft:myMycroft "Mycroft A.I." @ "Living Room" [host="192.168.X.X", port=8181] { 
+    Channels:
+        Type full-message-channel : Text [
+            messageTypes="message.type.1,message.type.4"
+        ]
+}
+```
+
+### Item Configuration
+
+The `mycroft.item` file :
+
+```java
+Switch myMycroft_mute                  "Mute"                      { channel="mycroft:mycroft:myMycroft:volume_mute" }
+Switch myMycroft_duck                  "Duck"                      { channel="mycroft:mycroft:myMycroft:volume_duck" }
+Dimmer myMycroft_volume                "Volume [%d]"               { channel="mycroft:mycroft:myMycroft:volume" }
+Player myMycroft_player                "Control"                   { channel="mycroft:mycroft:myMycroft:player" }
+Switch myMycroft_listen                "Wake and listen"           { channel="mycroft:mycroft:myMycroft:listen" }
+String myMycroft_speak                 "Speak STT"                 { channel="mycroft:mycroft:myMycroft:speak" }
+String myMycroft_utterance             "Utterance"                 { channel="mycroft:mycroft:myMycroft:utterance" }
+String myMycroft_fullmessage           "Full JSON message"         { channel="mycroft:mycroft:myMycroft:full_message" }
+```
+
+### Sitemap Configuration
+
+A `demo.sitemap` file :
+
+```perl
+sitemap demo label="myMycroft"
+{
+    Frame label="myMycroft" {
+        Switch    item=myMycroft_mute
+        Switch    item=myMycroft_duck
+        Slider    item=myMycroft_volume
+        Default   item=myMycroft_player
+        Switch    item=myMycroft_listen
+        Text      item=myMycroft_speak
+        Text      item=myMycroft_utterance
+        Text      item=myMycroft_fullmessage
+    }
+}
+```
+
+## Rule Actions
+
+This binding includes a number of rule actions, which allow triggering Mycroft from within rules.
+
+
+```
+(Rule DSL)
+val mycroftAction = getActions("mycroft","mycroft:mycroft:<uid>")
+```
+
+```
+(javascript JSR)
+var mycroftAction = actions.get("mycroft", "mycroft:mycroft:<uid>");
+```
+
+
+Where uid is the Thing UID of the Mycroft thing.
+
+Once this action instance is retrieved, you can invoke the `speak' method on it:
+
+```
+mycroftAction.speak("Hello world!")
+```
+
+
+The following actions are supported.
+
+
+| Action                     | Description  |
+|----------------------------|--------------|
+| speak(String message) | Mycroft will say this. |
```

Comment:
I think this is the right approach to have it as an "action", since its a one shot, i.e. fire&forget thing. But I am wondering why you have the `speak` channel defined as:
```
| speak                        | String    | The last sentence Mycroft speaks, or ask Mycroft to say something 
```
The `or ask Mycroft to say something` is a bit confusing to me. I would have expected it to be a `read-only` channel then.

Reply:
I have some difficulties to choose between about channel state / channel trigger / action.
I agree with you, speaking is a fire & forget thing. In fact it is not a state, as it is not maintained. It is "instantaneous".
Thus, it shouldn't even be a channel.
So I could make the TTS function a trigger channel (receive) AND an action (send).
But then, and it is where I'm lost, if I choose this kind of "bidirectionnal" functionnality, it seems logical to make it a state channel (as it handles very well the "bidirectionnal" use case).

(The same reasoning could be applied to the utterance channel)

I will trust your jugement on this and remove the "or ask mycroft to say something".
Should I also remove the "bidirectionnal" code on this channel, or should I keep it as an hidden functionnality in case someone is thinking "bidirectionnaly" like me ?

Reply:
Some clarification:

STATE channel:
Is a bidirectional channel that somehow persists a state. In case of your `speak channel` I would expect that if mycroft says something (i.e. something comes out of the speaker) this is represented on that channel. if that is the internal state of the speaker until it says something else, a STATE channel would be correct. So far it would be a read-only one though.
If you can define a sentence and make mycroft say this sentence AND mycroft stores this as its internal state too, THEN it makes sense to have it as a writeable state channel. However, in such a case the value on that state channel should always be in sync with the speakers internal state, i.e. if another 3rd party program would query the speaker about what he last said and this was triggered by openHAB the expectation would be that OH, the speaker and the 3rd party program will all have the same value.

TRIGGER channel:
This is a one directional channel, which is telling openHAB that something has happend and that "something" is usually well-known and one might even be able to enumerate the possible values. Usually this is used for remote controls where a button press like "long press", "short press", etc. will be modeled as such a channel. This is a fire&forget thing, there is no state involved neither on the device not on the openHAB side.

ThingAction:
This is the openHAB construct to model the opposite direction of a trigger channel. It is a fire&forget action to do something on a device which is not persisted in openHAB.

So for your `speak` and `utterance` channels it now depends on whether they change a state within the speaker which is persisted there and can be queried at any time or whether you just interact with it on a fire&forget basis.

Reply:
Mycroft doesn't keep the last message as an internal state. It is just a message on the application bus, listened once by the audio service (and by any other client listener such as this binding). So I guess by your definition it should not be a channel state on openHAB...
And as you say, a trigger channel is not really a good match, as the payload event should be ideally an element of an enumeration of well known values.
So, should it be removed ? But it would prevent openHAB to react on utterance or TTS event, which I think is an interesting possibility...

I looked at the Amazon Echo Control binding by curiosity and the same kind (approximately) of functionnality is implemented as channels (textToSpeech and lastVoiceCommand, the first one is write only and the second one is read/write), no Action.

It is so convenient to have it as a bidirectionnal item, that I am really at loss when deciding what to do.

My (personal) preference (driven unfortunately by my personal usage and not by coherence of concept, which has primauty) is to keep an Action, and also a bidirectionnal channel. But as I said earlier I'm OK with your decision. Just tell me what is best.

Reply:
>But it would prevent openHAB to react on utterance or TTS event, which I think is an interesting possibility...

No, we should definitely NOT remove this kind of functionality. I am just thinking about how to map it best to the OH architecture.

>I looked at the Amazon Echo Control binding by curiosity and the same kind (approximately) of functionnality is implemented as channels (textToSpeech and lastVoiceCommand, the first one is write only and the second one is read/write), no Action.

That might also be due to the "age" of that binding. the `ThingActions` were introduced into OH much later.

The only thing I am struggling with is that there is no real `state` involved, i.e. something that is somewhat persisted on the speaker an can be queried at any time.

@kaikreuzer Do you have any preference on how to handle these 2 channels: `speak` and `utterance`? IMHO as he said they don't reflect an internal state of the speaker, so a trigger channel with a longer text as payload might make sense here. In the opposite direction I would vote for the thing actions, as they are fire&forget actions.

Reply:
Difficult indeed. If I understand it correctly, both `speak` and `utterance` channels provide strings also when the user directly uses the MyCroft - and those are the events you'd like to react upon. I would say that "last utterance" can be indeed seen as a state - even if not stored on the device itself, but it can be helpful to have it in a state within openHAB, e.g. to also display it in a UI. 
As such an item can then also be used to send a text (as a command, which then would also directly become the state of the item), I would argue that there is no need for an additional action as it wouldn't bring any additional benefit.

Reply:
>I would say that "last utterance" can be indeed seen as a state - even if not stored on the device itself, but it can be helpful to have it in a state within openHAB, e.g. to also display it in a UI.

Showing it in the UI would be a good argument for a `state` channel for `utterance` indeed!

Question to @dalgwen: Does the same hold for the `speak` channel? I.e. if you interact with the speaker only, do you get that information about what it last said in openHAB? If so, I would also create a `state` channel for it. If the openHAB channel can only be filled by using openHAB to send a text to the speaker which should be said, then I would go for the thing action.

Reply:
The speaker and utterance channels works exactly in the same way.
The Mycroft binding connects to the Mycroft message bus (which is a publish / subscribe websocket). This bus is a a central point between Mycroft modules (mainly "skills", "audio", "voice")
By listening to every messages exchanged between modules, the binding can updates the relative channels in openHAB.
Example in our two cases : 
- when the voice module transmit an utterance spoken by the user to the skills module --> channel uterrance
- when the skills module ask the audio module to say something in the speaker --> channel speak
The two channels, in the opposite direction (when updated in openHAB), can inject a message in the bus to recreate the same behavior in Mycroft.

> Does the same hold for the `speak` channel? I.e. if you interact with the speaker only, do you get that information about what it last said in openHAB? 

So, to answer your question, yes, they are "two-ways" channels which can be updated even with no openHAB originating command.

Then, if I understand clearly, I guess the two-ways channels are a good match ?
I will keep them, and as you advise it, I will also delete the speak and utterance actions as they are redundant.

Reply:
Please do so, @dalgwen!